File: scala/bob/src/main/scala/Bob.scala
class Bob {

  def hey(input: String): String = {
    val default = "Whatever."
    val shouting = "Whoa, chill out!"
    val question = "Sure."
    val silence = "Fine. Be that way!"

    input match {
      case _ if isSilence(input) => silence
      case _ if isQuestion(input) => question
      case _ if isShouting(input) => shouting
      case _ => default
    }
  }

  def isShouting(input: String): Boolean = {
    isAllCaps(input) && hasLetters(input)
  }

  def isQuestion(input: String): Boolean = {
    ! hasLetters(input) && input.last == '?' ||
    input.last == '?' && ! isShouting(input)
  }

  def isSilence(input: String): Boolean = {
    input.trim.isEmpty
  }

  def isAllCaps(input: String): Boolean = {
    "[a-z]+".r.findFirstIn(input).isEmpty
  }

  def hasLetters(input: String): Boolean = {
    "[a-zA-Z]+".r.findFirstIn(input).isDefined
  }
}

File: haskell-old/bob/src/Bob.hs
module Bob
    ( responseFor
    ) where

import Data.Char

responseFor :: String -> String
responseFor xs
    | shoutingQuestion = "Calm down, I know what I'm doing!"
    | shouting = "Whoa, chill out!"
    | silence = "Fine. Be that way!"
    | question == '?' = "Sure."
    | otherwise = "Whatever."
  where
    shouting = (not $ null $ filter isAlpha xs) && (all isUpper $ filter isAlpha $ xs)
    shoutingQuestion = shouting && (last xs == '?')
    question = last (filter (not . isSpace) xs)
    silence = all isSpace xs
